src/client/pkg/grpcutil/pool.go

```diff
@@ -2,29 +2,85 @@ package grpcutil
 
 import (
 	"context"
+	"fmt"
+	"sync"
 
 	"google.golang.org/grpc"
+	"k8s.io/kubernetes/pkg/api"
+	kube "k8s.io/kubernetes/pkg/client/unversioned"
+	"k8s.io/kubernetes/pkg/labels"
+	"k8s.io/kubernetes/pkg/watch"
 )
 
 // Pool stores a pool of grpc connections to, it's useful in places where you
 // would otherwise need to create several connections.
 type Pool struct {
-	address string
-	opts    []grpc.DialOption
-	conns   chan *grpc.ClientConn
+	// The addresses of the pods that we currently know about
+	addresses []string
+	// The index to the next pod that we will dial
+	addressesIndex int
+	addressesLock  sync.Mutex
+	// addressesCond is used to notify goros that addresses have been obtained
+	addressesCond  *sync.Cond
+	endpointsWatch watch.Interface
+	opts           []grpc.DialOption
+	conns          chan *grpc.ClientConn
 }
 
-// NewPool creates a new connection pool, size is the maximum number of
-// connections that it will cache. There is no limit to he number of
-// connections that it can provide.
-func NewPool(address string, size int, opts ...grpc.DialOption) *Pool {
-	return &Pool{
-		address: address,
-		opts:    opts,
-		conns:   make(chan *grpc.ClientConn, size),
+// NewPool creates a new connection pool with connections to pods in the
+// given service.
+func NewPool(kubeClient *kube.Client, namespace string, serviceName string, numWorkers int, opts ...grpc.DialOption) (*Pool, error) {
+	endpointsInterface := kubeClient.Endpoints(namespace)
+
+	watch, err := endpointsInterface.Watch(api.ListOptions{
+		LabelSelector: labels.SelectorFromSet(
+			map[string]string{"app": serviceName},
+		),
+		Watch: true,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	pool := &Pool{
+		endpointsWatch: watch,
+		opts:           opts,
+		conns:          make(chan *grpc.ClientConn, numWorkers),
+	}
+	pool.addressesCond = sync.NewCond(&pool.addressesLock)
+	go pool.watchEndpoints()
+	return pool, nil
+}
+
+func (p *Pool) watchEndpoints() {
+	for {
+		event, ok := <-p.endpointsWatch.ResultChan()
+		if !ok {
+			return
+		}
+
+		endpoints := event.Object.(*api.Endpoints)
+		p.updateAddresses(endpoints)
 	}
 }
 
+func (p *Pool) updateAddresses(endpoints *api.Endpoints) {
+	var addresses []string
+	for _, subset := range endpoints.Subsets {
+		// According the k8s docs, the full set of endpoints is the cross
+		// product of (addresses x ports).
+		for _, address := range subset.Addresses {
+			for _, port := range subset.Ports {
+				addresses = append(addresses, fmt.Sprintf("%s:%d", address.IP, port.Port))
+			}
+		}
+	}
```

Comment:
Hmm, if `subset.Ports` has more than 1 port in it won't we use more than 1 port?